source/slang/slang-ir-any-value-marshalling.cpp

```diff
@@ -541,4 +651,97 @@ namespace Slang
         context.sharedContext = sharedContext;
         context.processModule();
     }
+
+    SlangInt alignUp(SlangInt x, SlangInt alignment)
+    {
+        return (x + alignment - 1) / alignment * alignment;
+    }
+
+    SlangInt _getAnyValueSizeRaw(IRType* type, SlangInt offset)
+    {
+        switch (type->getOp())
+        {
+        case kIROp_IntType:
+        case kIROp_FloatType:
+        case kIROp_UIntType:
+            return alignUp(offset, 4) + 4;
+        case kIROp_UInt64Type:
+        case kIROp_Int64Type:
+        case kIROp_DoubleType:
+            return -1;
+        case kIROp_Int16Type:
+        case kIROp_UInt16Type:
+        case kIROp_HalfType:
+            return alignUp(offset, 2) + 2;
+        case kIROp_UInt8Type:
+        case kIROp_Int8Type:
+            return -1;
+        case kIROp_VectorType:
+        {
+            auto vectorType = static_cast<IRVectorType*>(type);
+            auto elementType = vectorType->getElementType();
+            auto elementCount = getIntVal(vectorType->getElementCount());
+            for (IRIntegerValue i = 0; i < elementCount; i++)
+            {
+                offset = _getAnyValueSizeRaw(elementType, offset);
+                if (offset < 0) return offset;
+            }
+            return offset;
+        }
+        case kIROp_MatrixType:
+        {
+            auto matrixType = static_cast<IRMatrixType*>(type);
+            auto elementType = matrixType->getElementType();
+            auto colCount = getIntVal(matrixType->getColumnCount());
+            auto rowCount = getIntVal(matrixType->getRowCount());
+            for (IRIntegerValue i = 0; i < colCount; i++)
+            {
+                for (IRIntegerValue j = 0; j < rowCount; j++)
+                {
+                    offset = _getAnyValueSizeRaw(elementType, offset);
+                    if (offset < 0) return offset;
+                }
+            }
+            return offset;
+        }
+        case kIROp_StructType:
+        {
+            auto structType = cast<IRStructType>(type);
+            for (auto field : structType->getFields())
+            {
+                offset = _getAnyValueSizeRaw(field->getFieldType(), offset);
+                if (offset < 0) return offset;
+            }
+            return offset;
+        }
```

Comment:
Hmm... I'm a bit concerned if we end up having so many distinct pieces of layout logic that are *nearly* the same:

1. The AST-level layout rules for structured buffers
2. The IR-level type layout logic (implemented in `ir-type-layout`)
3. This logic, which defines a de facto distinct set of layout rules for "any-value" packing

The intention is that (2) is exactly compatible with (1) on all targets, but we don't currently test/enforce that (so it could be a source of bugs).

Looking at the code here, I'm worried because the struct case here isn't actually compatible with (2) or (1). It is going to be mostly the same, but then fail to match in a few cases, and I could see those creating issues for users.

This is especially notable because we don't have AST-level layout logic for this new format under (3), and that means that there is no way for a client of the reflection API (like `gfx`) to fill in a buffer using that layout.

I think this is a fine solution for now, but in the long run we really need the packing of data in an any-value to exactly match the layout for (1) and (2) above, so that we have a consistent notion of what "natural" layout is, and use that in as many cases as possible.

Reply:
Note that a very real problem with this kind of approach to structure layout is that by not computing an alignment for the whole struct and applying that, you end up in a situation where different fields using the same struct type may lay out differently because they started at differently aligned offsets.

This logic having to loop through all fields or all array elements also bloats the time complexity of this code. If it was about to rely on the existing IR type layout it could take advantage of the  way those memo-cache the layout information on the IR nodes to speed up subsequent queries.

Reply:
Right, the compiler behavior is overly complex due to this specific kind of layout. We should change them altogether to use (2) in a separate PR.